source/hash_table.c

```diff
@@ -0,0 +1,582 @@
+    if (state->entry_count + 1 > state->max_load) {
+        rv = expand_table(map);
+        if (rv != AWS_ERROR_SUCCESS) {
+            return rv;
+        }
+        state = map->pImpl;
+        // If we expanded the table, we need to discard the probe index returned from find_entry,
+        // as it's likely that we can find a more desirable slot. If we don't, then later gets will
+        // terminate before reaching our probe index.
+
+        // n.b. currently we ignore this probe_idx subsequently, but leaving this here so we don't
+        // forget when we optimize later.
+        probe_idx = 0;
+    }
+
+    state->entry_count++;
+    struct hash_table_entry new_entry;
+    new_entry.element.key = key;
+    new_entry.element.value = NULL;
+    new_entry.hash_code = hash_code;
+
+    entry = emplace_item(state, new_entry, probe_idx);
+
+    if (pElem) {
+        *pElem = &entry->element;
+    }
+
+    *was_created = 1;
+
+    return AWS_ERROR_SUCCESS;
+}
+
+// Clears an entry. Does _not_ invoke destructor callbacks.
+// Returns the last slot touched (note that if we wrap, we'll report an index lower than the
+// original entry's index)
+static int remove_entry(struct hash_table_state *state, struct hash_table_entry *entry) {
+    state->entry_count--;
+
+    // Shift subsequent entries back until we find an entry that belongs at its current position.
+    // This is important to ensure that subsequent searches don't terminate at the removed element.
+    int index = index_for(state, entry);
+    while (1) {
+        int next_index = (index + 1) & state->mask;
+
+        // If we hit an empty slot, stop
+        if (!state->slots[next_index].hash_code) break;
+        // If the next slot is at the start of the probe sequence, stop.
+        // We know that nothing with an earlier home slot is after this; otherwise
+        // this index-zero entry would have been evicted from its home.
+        if ((state->slots[next_index].hash_code & state->mask) == next_index) break;
+
+        // Okay, shift this one back
+        memcpy(&state->slots[index], &state->slots[next_index], sizeof(*state->slots));
+        index = next_index;
+    }
+
+    // Clear the entry we shifted out of
+    memset(&state->slots[index], 0, sizeof(*entry));
+
+    return index;
+}
+
+int aws_common_hash_table_remove(struct aws_common_hash_table *map,
+    const void *key, struct aws_common_hash_element *pValue
+) {
+    struct hash_table_state *state = map->pImpl;
+    uint64_t hash_code = hash_for(state, key);
+    struct hash_table_entry *entry;
+
+    int rv = find_entry(state, hash_code, key, &entry, NULL);
+
+    if (rv != AWS_ERROR_SUCCESS) {
+        return rv;
+    }
+
+    if (pValue) {
+        *pValue = entry->element;
+    } else if (state->destroy_fn) {
+        state->destroy_fn(entry->element);
+    }
+
+    remove_entry(state, entry);
+
+    return AWS_ERROR_SUCCESS;
+}
+
+int aws_common_hash_table_foreach(struct aws_common_hash_table *map,
```

Comment:
Maybe add in the API documentation, that any iteration ordering is purely an accidental result of the implementation and should not be relied upon, for any reason, ever.